src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -0,0 +1,206 @@
+        def trace_shifts(stencil: itir.Expr, inputs: list[itir.Expr], domain: itir.Expr):
+            node = itir.StencilClosure(
+                stencil=stencil,
+                inputs=inputs,
+                output=im.ref(out_field_name),
+                domain=domain,
+            )
+            return TraceShifts.apply(node)
+
+        # Extract the shifts and translate the domains accordingly
+        shifts_results = trace_shifts(stencil, inputs_node, SymbolicDomain.as_expr(symbolic_domain))
+
+        for in_field in inputs:
+            in_field_id = str(in_field.id)
+            shifts_list = shifts_results[in_field_id]
+
+            new_domains = [
+                cls._translate_domain(symbolic_domain, shift, offset_provider)
+                for shift in shifts_list
+            ]
+
+            accessed_domains[in_field_id] = domain_union(new_domains)
+
+        inputs_new = []
+        for in_field in inputs:
+            # Recursively traverse inputs
+            if isinstance(in_field, itir.FunCall):
+                transformed_calls_tmp, accessed_domains_tmp = cls.infer_as_fieldop(
+                    in_field, accessed_domains[str(in_field.id)], offset_provider
+                )
+                inputs_new.append(transformed_calls_tmp)
+
+                # Merge accessed_domains and accessed_domains_tmp
+                accessed_domains = cls._merge_domains(accessed_domains, accessed_domains_tmp)
+            else:
+                inputs_new.append(in_field)
+
+        transformed_call = im.call(
+            im.call("as_fieldop")(stencil, SymbolicDomain.as_expr(symbolic_domain))
+        )(*inputs_new)
+
+        accessed_domains_without_tmp = {
+            k: v for k, v in accessed_domains.items() if not k.startswith("__dom_inf_")
+        }
+
+        return transformed_call, accessed_domains_without_tmp
+
+    @staticmethod
+    def _validate_temporary_usage(body: list[itir.SetAt], temporaries: list[str]):
+        for i in range(len(body)):
+            for j in range(i + 1, len(body)):
+                if (
+                    str(body[i].target.id) == str(body[j].target.id)
+                    and str(body[i].target.id) in temporaries
+                ):
+                    raise ValueError("Temporaries can only be used once within a program.")
+
+    @classmethod
+    def infer_program(
+        cls,
+        program: itir.Program,
+        offset_provider: Dict[str, Dimension],
+    ) -> itir.Program:
+        fields_dict = {}
+        new_set_at_list = []
+
+        temporaries = [str(tmp.id) for tmp in program.declarations]
+
+        cls._validate_temporary_usage(program.body, temporaries)
+
+        for set_at in reversed(program.body):
+            assert isinstance(set_at, itir.SetAt)
+            assert isinstance(set_at.expr, itir.FunCall)
+            assert isinstance(set_at.expr.fun, itir.FunCall)
+            assert set_at.expr.fun.fun == im.ref("as_fieldop")
```

Comment:
I noticed that this is too restrictive for what we need. That was not clear when this project started. We now allow also pure field expressions at this level, e.g. 

```
__field_operator_testee(
  __sym_1,
  out,
  ____sym_1_size_0,
  ____sym_1_size_1,
  ____sym_1_size_2,
  __out_size_0,
  __out_size_1,
  __out_size_2
) {
  testee = λ(a) → a;
  out @ c⟨ IDimₕ: [0, __out_size_0), JDimₕ: [0, __out_size_1), KDimᵥ: [0, __out_size_2) ⟩
       ← testee(__sym_1);
}
```

Reply:
Implemented (in the let branch).